ebpf/monitor/accesslog.proto

```diff
@@ -0,0 +1,340 @@
+  // total local out use duration(nanoseconds) -> kernel: ip_local_out
+  uint64 totalLocalDuration = 2;
+  // total output use duration(nanoseconds) -> kernel: ip_finish_output2 - ip_output
+  uint64 totalOutputDuration = 3;
+  // total resolve remote MAC address(ARP Request) count and duration(nanosecond) -> kernel: neigh_resolve_output
+  uint64 totalResolveMACCount = 5;
+  uint64 totalResolveMACDuration = 6;
+  // total netfiltering count and duration(nanosecond) -> kernel: nf_hook
+  uint64 totalNetFilterCount = 7;
+  uint64 totalNetFilterDuration = 8;
+}
+
+
+message AccessLogKernelWriteL2Metrics {
+  // total duration(nanosecond) of layer 2
+  uint64 totalDuration = 1;
+  // target network hardware interface index, get the net hardware name through EBPFAccessLogNodeNetInterface#index
+  uint32 ifindex = 2;
+  // total enter the net device buffer count
+  uint64 totalEnterQueueBufferCount = 3;
+  // total buffer data ready to send duration(nanosecond), ready_to_send - enter_queue
+  uint64 totalReadySendDuration = 4;
+  // total send buffer to the net device duration(nanosecond)
+  uint64 totalNetDeviceSendDuration = 5;
+}
+
+message AccessLogKernelReadOperation {
+  // Starting to read data timestamp
+  EBPFTimestamp startTime = 1;
+  // Finish read operation timestamp
+  EBPFTimestamp endTime = 2;
+  // Which kind of syscall of current read operation
+  AccessLogKernelReadSyscall syscall = 3;
+
+  // Layer 2-4 related metrics
+  AccessLogKernelReadL2Metrics l2Metrics = 4;
+  AccessLogKernelReadL3Metrics l3Metrics = 5;
+  AccessLogKernelReadL4Metrics l4Metrics = 6;
+}
+
+message AccessLogKernelReadL2Metrics {
+  // from network hardware interface index, get the net hardware name through EBPFAccessLogNodeNetInterface#index
+  uint32 ifindex = 1;
+  // total package count
+  uint32 totalPackageCount = 2;
+  // total package size
+  uint64 totalPackageSize = 3;
+  // total package from receive to enqueue(wait for reading) duration(nanosecond)
+  uint64 totalPackageToQueueDuration = 4;
+}
+
+message AccessLogKernelReadL3Metrics {
+  uint64 totalDuration = 1;
+  // total local receive use duration(nanoseconds) -> kernel: ip_rcv_finish - ip_rcv
+  uint64 totalRecvDuration = 2;
+  // total local use duration(nanoseconds) -> kernel: ip_local_deliver_finish - ip_local_deliver
+  uint64 totalLocalDuration = 3;
+  // total netfiltering count and duration(nanosecond) -> kernel: nf_hook
+  uint64 totalNetFilterCount = 4;
+  uint64 totalNetFilterDuration = 5;
+}
+
+message AccessLogKernelReadL4Metrics {
+  uint64 totalDuration = 1;
+}
+
+enum AccessLogKernelWriteSyscall {
+  Write = 0;
+  Writev = 1;
+  Send = 2;
+  SendTo = 3;
+  SendMsg = 4;
+  SendMmsg = 5;
+  SendFile = 6;
+  SendFile64 = 7;
+}
+
+enum AccessLogKernelReadSyscall {
+  Read = 0;
+  Readv = 1;
+  Recv = 2;
+  RecvFrom = 3;
+  RecvMsg = 4;
+  RecvMmsg = 5;
+}
+
+message EBPFTimestamp {
+  oneof timestamp {
+    EBPFOffsetTimestamp offset = 1;
+    Instant absolute = 2;
+  }
```

Comment:
Could you explain how this timestamp works?

Reply:
In the kernel, we can get different timestamps. In the offset mode, we can only get the timestamp offset(through `bpf_ktime_get_ns`), and then we need to add it with the system boot time to get the real-time, that's why I added the system boot time in the node info. But in some kernel methods, we can only get the absolute timestamp.

Reply:
Is the timestamp sent as always in a bulk at least?